packages/flutter_tools/lib/src/commands/clean.dart

```diff
@@ -28,38 +32,69 @@ class CleanCommand extends FlutterCommand {
 
   @override
   Future<FlutterCommandResult> runCommand() async {
+    // Clean Xcode to remove intermediate DerivedData artifacts.
+    // Do this before removing ephemeral directory, which would delete the xcworkspace.
+    final FlutterProject flutterProject = FlutterProject.current();
+    if (xcode.isInstalledAndMeetsVersionCheck) {
+      await _cleanXcode(flutterProject.ios);
+      await _cleanXcode(flutterProject.macos);
+    }
+
     final Directory buildDir = fs.directory(getBuildDirectory());
-    _deleteFile(buildDir);
+    deleteFile(buildDir);
 
-    final FlutterProject flutterProject = FlutterProject.current();
-    _deleteFile(flutterProject.dartTool);
+    deleteFile(flutterProject.dartTool);
 
     final Directory androidEphemeralDirectory = flutterProject.android.ephemeralDirectory;
-    _deleteFile(androidEphemeralDirectory);
+    deleteFile(androidEphemeralDirectory);
 
     final Directory iosEphemeralDirectory = flutterProject.ios.ephemeralDirectory;
-    _deleteFile(iosEphemeralDirectory);
+    deleteFile(iosEphemeralDirectory);
+
+    final Directory macosEphemeralDirectory = flutterProject.macos.ephemeralDirectory;
+    deleteFile(macosEphemeralDirectory);
 
     return const FlutterCommandResult(ExitStatus.success);
   }
 
-  void _deleteFile(FileSystemEntity file) {
-    final String path = file.path;
-    printStatus("Deleting '$path${fs.path.separator}'.");
+  Future<void> _cleanXcode(XcodeBasedProject xcodeProject) async {
+    if (xcodeProject.existsSync()) {
```

Comment:
Could reduce the nesting here with an early return.

```suggestion
    if (!xcodeProject.existsSync()) {
      return;
    }
```